traversal/scanner/GraphIterator.java

```diff
@@ -377,18 +377,70 @@ private void reset() {
         }
 
         private Forwardable<Vertex<?, ?>, Order.Asc> createIteratorFromEdges() {
-            List<Forwardable<Vertex<?, ?>, Order.Asc>> iters = new ArrayList<>();
-            edgeInputs.forEach((procedureEdge, vertex) -> iters.add(branch(vertex, procedureEdge)));
-            if (iters.size() == 1) return iters.get(0);
-            else return intersect(iterate(iters), ASC);
+            Map<Vertex<?, ?>, List<ProcedureEdge.Native.Thing.RolePlayer>> overlappingRPEdges = new HashMap<>();
+            List<ProcedureEdge<?, ?>> otherEdges = new ArrayList<>();
+            edgeInputs.forEach((edge, vertex) -> {
+                if (edge.isRolePlayer()) {
+                    overlappingRPEdges.computeIfAbsent(vertex, v -> new ArrayList<>()).add(edge.asRolePlayer());
+                } else otherEdges.add(edge);
+            });
+            return intersectEdges(overlappingRPEdges, otherEdges);
+        }
+
+        private Forwardable<Vertex<?, ?>, Order.Asc> intersectEdges(Map<Vertex<?, ?>, List<ProcedureEdge.Native.Thing.RolePlayer>> rpEdges,
+                                                                    List<ProcedureEdge<?, ?>> otherEdges) {
+            List<Forwardable<Vertex<?, ?>, Order.Asc>> iterators = new ArrayList<>();
+            otherEdges.forEach(e -> iterators.add(branch(edgeInputs.get(e), e)));
+            rpEdges.forEach((vertex, edges) -> {
+                if (edges.size() == 1) {
+                    iterators.add(branch(edgeInputs.get(edges.get(0)), edges.get(0))
+                            .mapSorted(KeyValue::key, thing -> KeyValue.of(thing.asThing(), null), ASC));
+                } else iterators.add(intersectOverlapping(edges));
+            });
```

Comment:
normal edges or role player edges that are not parallel are simply added to the list of iterators. If there are multiple parallel role player edges, we handle them separately